### PR TITLE
Fix LFS Locks over SSH

### LINKED ISSUE #3084: Git LFS Lock: credential issue when using SSH
- Gitea version (or commit ref): master (c0a7899416af728634b2dc52a27b31255285ae7e)
- Git version: 2.11.0
- Operating system: Ubuntu 17.06 / Debian 9
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes, but it is not exactly the same bug... https://try.gitea.io/testzzx464c/lfs-lock-test
  - [ ] No
  - [ ] Not relevant

## Description

Hello,

I am testing the new git LFS lock feature and I have some issues.

When I try it on my own machine and server, I have the following error:

```
$ git push
Username for 'http://localhost:3000': test
Password for 'http://test@localhost:3000': 
WARNING: Authentication error: Authentication required: You must have push access to verify locks : User undefined doesn't have rigth to verify for lfs lock [rid: 1]
[...]

$ git lfs locks
Error while retrieving locks: You must have pull access to list locks : User undefined doesn't have rigth to list for lfs lock [rid: 1]
```

When I try on try.gitea.io, I have no error on push but I cannot get lock info:

```
$ git lfs lock foo.yy 
Lock failed: Repository or object not found: https://try.gitea.io/testzzx464c/lfs-lock-test.git/info/lfs/locks
Check that it exists and that you have proper access to it
```

I am probably doing it wrong, if someone can tell me where... :)

cc @sapk

### PR DESCRIPTION
**Pull request by zeripath**
Whilst looking at adding more tests for #6993 I've noticed that LFS locking was not-preauthenticating  correctly over SSH.

This PR fixes this.

Fixes #3084

### CODE REVIEW

**Comment by lunny** [comment]
It seems there is an issue reported that.

**Comment by zeripath** [comment]
@lunny do you know which issue it is because I can't seem to easily find it.

**Comment by lunny** [comment]
Maybe it is https://github.com/go-gitea/gitea/issues/5478

**Comment by zeripath** [comment]
It's possible that this is part of the fix for that, but I think #6993 is probably more likely to fix that.

I'm going to incorporate this in #6993 in any case.

**Review by lafriks — APPROVED** [review]

File: integrations/git_test.go [hunk 1]
```diff
@@ -197,10 +212,17 @@ func testGit(t *testing.T, u *url.URL) {
 					t.Run("Little", func(t *testing.T) {
 						PrintCurrentTest(t)
 						littleLFS = commitAndPush(t, littleSize, dstPath)
+						lockFileTest(t, littleLFS, dstPath)
+
 					})
 					t.Run("Big", func(t *testing.T) {
+						if testing.Short() {
+							return
```
Comment (sapk) [inline integrations/git_test.go:223]: Please use t.Skip method to indicate we skipped the test.
        
```suggestion
							t.Skip("skipping test in short mode.")
```

**Review by sapk — CHANGES_REQUESTED** [review]
You need to use the Skip method in place of simple return in skipped tests. Overall looks good but I haven't tested yet.

File: integrations/git_test.go (hunk 1 above)
Comment (zeripath) [reply integrations/git_test.go:223]: Done

File: integrations/git_test.go [hunk 2]
```diff
@@ -65,6 +65,9 @@ func testGit(t *testing.T, u *url.URL) {
 					little = commitAndPush(t, littleSize, dstPath)
 				})
 				t.Run("Big", func(t *testing.T) {
+					if testing.Short() {
+						return
```
Comment (sapk) [inline integrations/git_test.go:70]: Missing also a skip here.

Comment (zeripath) [reply integrations/git_test.go:70]:
```suggestion
						t.Skip("skipping test in short mode.")
						return
```

Comment (zeripath) [reply integrations/git_test.go:70]: Done

File: integrations/git_test.go (hunk 1 above)
Comment (sapk) [reply integrations/git_test.go:223]: In fact this was not ...

**Comment by zeripath** [comment]
@sapk have you been able to test this yet? git_test.go does test the locking now btw.

File: integrations/git_test.go [hunk 3]
```diff
@@ -268,15 +296,17 @@ func ensureAnonymousClone(t *testing.T, u *url.URL) {
 }
 
 func lockTest(t *testing.T, remote, repoPath string) {
-	_, err := git.NewCommand("remote").AddArguments("set-url", "origin", remote).RunInDir(repoPath) //TODO add test ssh git-lfs-creds
```
Comment (lunny) [inline integrations/git_test.go:271]: So we ignored the `remote` parameter?

File: routers/routes/routes.go [hunk 4]
```diff
@@ -923,7 +923,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 					m.Post("/", lfs.PostLockHandler)
 					m.Post("/verify", lfs.VerifyLockHandler)
 					m.Post("/:lid/unlock", lfs.UnLockHandler)
-				}, context.RepoAssignment())
```
Comment (lunny) [inline routers/routes/routes.go:926]: Why we remove this?

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/routes/routes.go (hunk 4 above)
Comment (zeripath) [reply routers/routes/routes.go:926]: LFS has a different authentication scheme to that of the normal repos. RepoAssignment simply doesn't work for it.

File: integrations/git_test.go (hunk 3 above)
Comment (zeripath) [reply integrations/git_test.go:271]: It's not necessary - the remote parameter is set already when we clone.

This line is incorrect for SSH cloned repos

Comment (zeripath) [reply integrations/git_test.go:271]: I should say that this line is what was hiding the SSH locking bug from being detected. It's incorrectly overriding the remote that is being set on clone .

File: routers/routes/routes.go (hunk 4 above)
Comment (zeripath) [reply routers/routes/routes.go:926]: Yes, we could/should make LFS's authentication a macaron handler but we should do that as part of a larger refactor of the LFS server - along with a check that we are not missing functionality from git-lfs's lfstest-gitserver.

But that's another pr.

**Comment by sapk** [comment]
@zeripath not much time to do so lately. So if someone other validate it should be good as my requested changes are done.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@zeripath  I think this should resolve #7219, could you send back port to v1.8 .

**Pull request merged**